File: Leetcode/Array/683_K_Empty_Slots.py
class Solution(object):
    def kEmptySlots(self, flowers, k):
        """
        :type flowers: List[int]
        :type k: int
        :rtype: int
        """
        f=[False]*len(flowers)

        def isValid(bloom_index, f,k):
        	f[bloom_index]=True

        	k_empty_right=True

        	for i in range(bloom_index+1,bloom_index+k+1):	
        		if i>=len(f) or f[i]:
        			k_empty_right=False
        			break
        	if k_empty_right and bloom_index+k+1<len(f) and f[bloom_index+k+1]:
        		return True

        	k_empty_left=True
        	for i in range(bloom_index-k,bloom_index):
        		if i<0 or f[i]:
        			k_empty_left=False
        			break
        	
        	if k_empty_left and bloom_index-k-1>=0 and f[bloom_index-k-1]:
        		return True

        	
        	
        	return False


        for i in range(len(flowers)):
        	if isValid(flowers[i]-1,f,k):
        		return i+1
        return -1

print Solution().kEmptySlots([1,3,2],1)
print Solution().kEmptySlots([1,2,3],1)

File: Leetcode/Array/683_K_Empty_Slots.cpp
#include <vector>
#include <set>
using namespace std;

class Solution {
public:
    int kEmptySlots(vector<int>& flowers, int k) {
        int n=flowers.size();
        if(n==0 || k>=n) return -1;

        set<int> xs;
    	for (int i=0;i<n;i++)
    	{
    		int x=flowers[i];
    		auto r=xs.insert(x).first;
    		auto l=r;
    		if (++r!=xs.end() && *r == x+k+1) return i+1;
    		if (l!=xs.begin() && *(--l)==x-k-1) return i+1;
    	}

    	return -1;
    }
};

class Solution2 {
public:
	int kEmptySlots(vector<int>& flowers, int k){
		int n=flowers.size();
		if (n==0 || k>n) return -1;
		++k;
		int bs=(n+k-1)/k;
		vector<int> lows(bs, INT_MAX);
		vector<int> highs(bs,INT_MIN);
		for (int i=0;i<n;i++){
			int x=flowers[i];
			int p=x/k;
			if (x<lows[p]){
				lows[p]=x;
				if(p>0 && highs[p-1]==x-k) return i+1;
			}
			if (x>highs[p]){
				highs[p]=x;
				if(p< bs-1 && lows[p+1]==x+k) return i+1;
			}
		}
		return -1;
	}	
};

int main()
{
	Solution s1;
	Solution2 s2;
}